### DISCUSSION TITLE
Trying the small example programs

### DISCUSSION THREAD

**Discussion by HM0880** [coalton-lang/coalton#192] [Help!] [answered]
## Startup

This is misc feedback on trying the small example programs. I am running the code from the directory `~/quicklisp/local-projects/coalton/examples/small-coalton-programs/` with a new instance of SLIME (v2.26.1).

`(ql:quickload :small-coalton-programs)` is successful.

## Running `(microbench1:run)` from SLIME

`(microbench1:run)` is successful.

    1  CL-USER> (microbench1:run)
    2  Lisp took 17374 ms
    3  Coalton took 22014 ms
    4  Lisp was faster by 4640 ms (21.08%)
    5  The hand-optimized function took 4379 ms
    6  NIL
    7  CL-USER>

### Unbound-variable error when running the entire file

Running `C-x h` (mark-whole-buffer) then `C-c C-c` (slime-compile-defun) gave the error, "The variable \#:SMALL-COALTON-PROGRAMS is unbound. [Condition of type COMMON-LISP:UNBOUND-VARIABLE]".

I do not know why the `(in-package #:small-coalton-programs)` gives this error.

### Output when running selected regions of the file

I bypassed the unbound-variable error by selecting line 7 to line 38 (i.e., the first appearance of `coalton-toplevel`) and running `C-c C-c`. This gave the output "; processing (COALTON-TOPLEVEL (DEFINE-TYPE SYMBOL &#x2026;) &#x2026;)" in SLIME, which seems right.

Selecting the last four lines and running `C-c C-c` gave "; processing (COALTON-TOPLEVEL (DEFINE TWO-X &#x2026;))". There are no errors, but I was expecting to see something like "2\*x" since the input is "x\*x", which is "x<sup>2</sup>". **What should I expect to see?**

## Running `diff.lisp` from the SBCL REPL

I tried running `diff.lisp` directly from the SBCL REPL. Running `(load "diff.lisp")` successfully finished with `T`.

**Is Coalton meant to be run only from the SBCL REPL and not from SLIME?**

     1  $ sbcl
     2  This is SBCL 2.1.9, an implementation of ANSI Common Lisp.
     3  [...]
     4  
     5  > (ql:quickload :small-coalton-programs)
     6  To load "small-coalton-programs":
     7    Load 1 ASDF system:
     8      small-coalton-programs
     9  ; Loading "small-coalton-programs"
    10  ....
    11  (:SMALL-COALTON-PROGRAMS)
    12  
    13  > (load "diff.lisp")
    14  T

**Comment by stylewarning** [comment]
There are dependencies. For example, you need to have the package before you can evaluate the remainder of the file.

Coalton should definitely work with SLIME fine. 

I recommend using `C-c C-l` to load entire files at once in SLIME, and not highlighting stuff and doing `C-c C-c`.

**Reply by HM0880** [reply]
Thanks.  Even without the `cl:` prefix, loading the file with `C-c C-l` was successful.  TIL...

**Comment by eliaslfox** [comment, marked as answer]
I think the in-package definition was broken because it wasn't prefixed correctly. This will be fixed by #193. 

With that change I was able to run the whole file `C-c C-l` and also run the individual blocks `C-c C-c` without issue.

**Reply by stylewarning** [reply]
Ah, good point. This depends on what `*package*` was set to. In ordinary circumstances it shouldn't be an issue but it's pretty easy for that to get changed during development.

**Comment by eliaslfox** [comment]
@HM0880 I'm marking this as answered but if you still have issues please reopen it.

**Reply by HM0880** [reply]
@eliaslfox Yes, adding the `cl:` prefix made both the `C-c C-l` and the `C-c C-c` commands work as expected.  Thanks!
